support-workers/src/main/scala/com/gu/support/workers/lambdas/CreateZuoraSubscription.scala

```diff
@@ -263,32 +276,48 @@ object DigitalSubscriptionGiftRedemption {
 
   def redeemGift(
     redemptionData: RedemptionData,
-    gifteeUserId: String,
     requestInfo: RequestInfo,
     state: CreateZuoraSubscriptionState,
     zuoraService: ZuoraService,
     catalogService: CatalogService
   ): Future[HandlerResult[SendThankYouEmailState]] =
-    for {
-      giftSubscriptionFields <- zuoraService.getSubscriptionFromRedemptionCode(redemptionData.redemptionCode)
-      validatedSubscription <- Future.fromTry(validateRedemptionData(giftSubscriptionFields))
-      fullGiftSubscription <- zuoraService.getSubscriptionById(validatedSubscription.id)
-      newTermLength <- Future.fromTry(calculateNewTermLength(fullGiftSubscription, catalogService))
-      updateDataResponse <- zuoraService.updateSubscriptionRedemptionData(validatedSubscription.id, gifteeUserId, newTermLength)
-      handlerResult <- Future.fromTry(buildHandlerResult(updateDataResponse, state, redemptionData, requestInfo))
-    } yield handlerResult
-
-  def validateRedemptionData(existingSub: SubscriptionRedemptionQueryResponse) = {
+    zuoraService.getSubscriptionFromRedemptionCode(redemptionData.redemptionCode).flatMap(
+      redemptionQueryResponse =>
+        getSubscriptionState(redemptionQueryResponse, state.requestId.toString) match {
+          case u: Unredeemed => doRedemption(u.subscriptionId, state, redemptionData, requestInfo, zuoraService, catalogService)
+          case RedeemedInThisRequest => Future.fromTry(buildHandlerResult(UpdateRedemptionDataResponse(true), state, redemptionData, requestInfo))
+          case Redeemed => Future.fromTry(Failure(new RuntimeException(GetCodeStatus.CodeAlreadyUsed.clientCode)))
+          case NotFound => Future.fromTry(Failure(new RuntimeException(GetCodeStatus.NoSuchCode.clientCode)))
+        }
+    )
+
+  private def getSubscriptionState(existingSub: SubscriptionRedemptionQueryResponse, requestId: String) = {
     existingSub.records.headOption.map(
       existingSubFields =>
         if (existingSubFields.gifteeIdentityId.isEmpty)
-          Success(existingSubFields)
-        else
-          Failure(new RuntimeException(GetCodeStatus.CodeAlreadyUsed.clientCode))
-    ).getOrElse(Failure(new RuntimeException(GetCodeStatus.NoSuchCode.clientCode)))
+          Unredeemed(existingSubFields.id)
+        else if (existingSubFields.createdRequestId == requestId)
+          RedeemedInThisRequest
+        else Redeemed
+    ).getOrElse(NotFound)
   }
 
-  def calculateNewTermLength(subscription: Subscription, catalogService: CatalogService) = {
+  private def doRedemption(
```

Comment:
we have a redeemGift and a doRedemption, which don't have any meaningful distinction in the name.
The difference in the function is that the first one does checks first, wheras the second one just goes the actual steps.
Perhaps we could call them tryToRedeem and redeemInZuora